src/tests/Interop/TypeMap/TypeMapApp.cs

```diff
@@ -0,0 +1,231 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+extern alias Original;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Xunit;
+
+using DupType_MapObject = Original.Lib.AliasedName;
+using DupType_MapString = Lib.AliasedName;
+
+[assembly: TypeMapAssemblyTarget<MultipleTypeMapAssemblies>("TypeMapLib3")]
+[assembly: TypeMapAssemblyTarget<MultipleTypeMapAssemblies>("TypeMapLib4")]
+
+[assembly: TypeMapAssemblyTarget<UnknownAssemblyReference>("DoesNotExist")]
+
+[assembly: TypeMap<TypicalUseCase>("1", typeof(C1))]
+[assembly: TypeMap<TypicalUseCase>("2", typeof(S1))]
+[assembly: TypeMap<TypicalUseCase>("3", typeof(Guid))]
+[assembly: TypeMap<TypicalUseCase>("4", typeof(object))]
+
+[assembly: TypeMap<object>("1", typeof(string))]
+[assembly: TypeMap<string>("1", typeof(string))]
+[assembly: TypeMap<int[]>("1", typeof(string))]
+[assembly: TypeMap<C1.I1>("1", typeof(string))]
+[assembly: TypeMap<C1.I2<int>>("1", typeof(string))]
+[assembly: TypeMap<C1.I2<string>>("1", typeof(string))]
+[assembly: TypeMap<C2<int>.I1>("1", typeof(string))]
+[assembly: TypeMap<C2<int>.I2<int>>("1", typeof(string))]
+[assembly: TypeMap<C2<string>.I1>("1", typeof(string))]
+[assembly: TypeMap<C2<string>.I2<string>>("1", typeof(string))]
+
+[assembly: TypeMapAssociation<TypicalUseCase>(typeof(C1), typeof(C1))]
+[assembly: TypeMapAssociation<TypicalUseCase>(typeof(S1), typeof(S1))]
+[assembly: TypeMapAssociation<TypicalUseCase>(typeof(Guid), typeof(C1))]
+[assembly: TypeMapAssociation<TypicalUseCase>(typeof(string), typeof(S1))]
+
+[assembly: TypeMapAssociation<object>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<string>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<int[]>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C1.I1>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C1.I2<int>>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C1.I2<string>>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C2<int>.I1>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C2<int>.I2<int>>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C2<string>.I1>(typeof(object), typeof(string))]
+[assembly: TypeMapAssociation<C2<string>.I2<string>>(typeof(object), typeof(string))]
```

Comment:
More interesting key values: 
- Array type e.g. `int[]`
- Uninstantiated generic types e.g. `List<>`
- Instantiated generic types, e.g. `Dictionary<string, string>`

Reply:
I already have an array type and a instantiated generic type, are you talking about using these as the generic type itself?

As in:

`C2<int[]>` and `C2<C1<int>>`?

> Uninstantiated generic types e.g. List<>

These aren't permitted in attributes.

Reply:
I meant the types used as the keys in the map, not the types used as the universe IDs. I think all of the following should be possible:

```
[assembly: TypeMapAssociation<object>(typeof(List<int>), typeof(string))]
[assembly: TypeMapAssociation<object>(typeof(List<>), typeof(string))]
[assembly: TypeMapAssociation<object>(typeof(int[]), typeof(string))]
[assembly: TypeMapAssociation<object>(typeof(int*), typeof(string))]
```